File: python/flintdb/query.py
class Query:
    """Manages data querying.

    Attributes:
        _query (dict): The query content.
        _cache (bool): To allow caching query result.
        _ready (bool): To detect whether query content has been normalized.
        _table (Table): The table to query data from.
        _database (Database): The parent database of table.
        _limit (list): To slice query result.
    """
    def __init__(self, database: "Database") -> None:
        """Initializes the Query object.

        Args:
            database: The parent database instance.
        """
        self._query = {}
        self._cache = True
        self._ready = False
        self._table = None
        self._database = database
        self._limit = [0, None]

    def use_table(self, table_name: str) -> "Self":
        """Sets the table to query data from.

        Args:
            table_name: The table to query from.
        """
        self._table = self._database.table(table_name)
        return self

    def join(self, table_name: str, on: list, prefix: str | None = None) -> "Self":
        """Performs left outer join.

        Args:
            table_name: The name of the table to join.
            on: The join clause.
            prefix: The prefix for right row columns.
        """
        if "join" not in self._query:
            self._query["join"] = {}

        prefix = table_name + "." if prefix is None else prefix
        self._query["join"][table_name] = [on, prefix]
        return self

    def map(self, callback: "Callable") -> "Self":
        """Maps callback to rows.

        Args:
            callback: The callback to pass rows into.
        """
        if "map" not in self._query:
            self._query["map"] = []

        self._query["map"].append([callback, callback.__doc__])
        return self

    def where(self, column_name: str, operator: str, value: "Any") -> "Self":
        """Conditional filtering.

        Args:
            column_name: The column name to filter.
            operator: The operator sign.
            value: The value to compare.
        """
        if "where" not in self._query:
            self._query["where"] = {}

        self._query["where"][column_name] = [operator, value]
        return self

    def select(self, column_name: str, new_name: str) -> "Self":
        """Renames column.

        Args:
            column_name: The column's original name.
            new_name: The target name.
        """
        if "select" not in self._query:
            self._query["select"] = {}

        self._query["select"][column_name] = new_name
        return self

    def distinct(self, column_name: str) -> "Self":
        """Retrieves distinct, non repeated values.

        Args:
            column_name: The column to filter.
        """
        if "distinct" not in self._query:
            self._query["distinct"] = []

        self._query["distinct"].append(column_name)
        return self

    def sort(self, column_name: str, order: str = "ASC") -> "Self":
        """Sorts rows base on column's value.

        Args:
            column_name: The column to sort.
            order: The order in which to sort.

        Raises:
            ValueError: If order is not ASC or DESC.
        """
        order = order.upper()
        if order not in ("ASC", "DESC"):
            raise ValueError("Only \"ASC\" and \"DESC\" are valid orders")

        if "sort" not in self._query:
            self._query["sort"] = []

        order = True if order == "ASC" else False
        self._query["sort"].append([column_name, order])
        return self

    def filter(self, callback: "Callable") -> "Self":
        """Conditional filtering by mapping callback.

        Args:
            callback: The callback to pass rows into.
        """
        if "filter" not in self._query:
            self._query["filter"] = []

        self._query["filter"].append([callback, callback.__doc__])
        return self

    def limit(self, max: int, offset: int = 0) -> "Self":
        """Restrict the number of rows returned.

        Args:
            max: The maximum number of rows to retrieve.
            offset: The starting point for retrieving rows.

        Raises:
            ValueError: If max is less than 1.
        """
        if max < 1:
            raise ValueError("Maximum limit must be greater than zero")

        self._limit = [offset, max]
        return self

    def no_cache(self) -> "Self":
        """Disables query result caching."""
        self._cache = False
        return self

    def fetch(self) -> list:
        """Returns query result."""
        self.normalize()
        if self._cache:
            from .cache import Cache
            cache = Cache(*self.identifier())
            if cache.valid():
                limit_start = self._limit[0]
                limit_stop = self._limit[1]
                return cache.get()[limit_start:limit_stop]

        data = []
        for row in self._table.rows():
            if self._query["join"]:
                join_object = Join(row)
                for table_name, options in self._query["join"]:
                    join_object.table(self._database.table(table_name))
                    join_object.options(options[0])
                    join_row = join_object.row()
                    if join_row:
                        for key, value in join_row.columns():
                            key = options[1] + key
                            row[key] = value

            if self._query["map"]:
                for map in self._query["map"]:
                    map[0](row)

            if self._query["where"]:
                where_object = Where(row)
                for column, criteria in self._query["where"]:
                    where_object.criteria(column, *criteria)
                    if where_object.match():
                        data.append(row)
            else:
                data.append(row)

            if row not in data:
                continue

            for column_name, new_name in self._query["select"]:
                row.rename_column(column_name, new_name)

        for column in self._query["distinct"]:
            values = set(())
            data[:] = [
                row for row in data
                if row[column] not in values
                and not values.add(row[column])
            ]

        for sort in self._query["sort"]:
            try:
                data.sort(key=lambda row: row[sort[0]], reverse=sort[1])
            except TypeError:
                pass

        for _filter in self._query["filter"]:
            data = list(filter(_filter[0], data))

        if self._cache and data:
            cache.put(data)

        limit_start = self._limit[0]
        limit_stop = self._limit[1]
        return data[limit_start:limit_stop]

    def normalize(self) -> None:
        """Normalizes query content.

        Raises:
            ValueError: If table not provided.
        """
        if self._table is None:
            raise ValueError("Table must be specified")

        if self._ready:
            return

        if "join" not in self._query:
            self._query["join"] = {}
        self._query["join"] = sorted(self._query["join"].items())

        if "map" not in self._query:
            self._query["map"] = []
        self._query["map"].sort()

        if "where" not in self._query:
            self._query["where"] = {}
        self._query["where"] = sorted(self._query["where"].items())

        if "select" not in self._query:
            self._query["select"] = {}
        self._query["select"] = sorted(self._query["select"].items())

        if "distinct" not in self._query:
            self._query["distinct"] = []
        self._query["distinct"].sort()

        if "sort" not in self._query:
            self._query["sort"] = []
        self._query["sort"].sort()

        if "filter" not in self._query:
            self._query["filter"] = []
        self._query["filter"].sort()

        self._query = dict(sorted(self._query.items()))
        self._ready = True

    def identifier(self) -> list:
        """Returns table and query content."""
        self.normalize()
        table = self._table
        query = self._query
        return [table, query]

    def clear(self) -> "Self":
        """Resets query instance."""
        self._query = {}
        self._cache = True
        self._ready = False
        self._limit = [0, None]
        return self


class Join:
    """Manages table joins.

    Attributes:
        _table (Table): The table to join.
        _options (list): The join clause.
        _where (Where): The instance of Where object.
    """
    def __init__(self, row: "Row") -> None:
        """Initializes Join object.

        Args:
            row: The row from left table.
        """
        self._table = None
        self._options = None
        self._where = Where(row)

    def table(self, table: "Table") -> None:
        """Sets the table to join.

        Args:
            table: The table to join.
        """
        self._table = table

    def options(self, options: list) -> None:
        """Sets the join clause.

        Args:
            options: The join clause.
        """
        self._options = options

    def filter(self, row: "Row") -> bool:
        """Filter rows.

        Args:
            row: The row from right table.

        Returns:
            True if it matches.
        """
        left_column = self._options[0]
        operator = self._options[1]
        right_column = self._options[2]
        value = row[right_column]

        self._where.criteria(left_column, operator, value)
        return self._where.match()

    def row(self) -> "Any":
        """Returns matching row or None."""
        query = self._table.query()
        query.no_cache()
        query.filter(self.filter)
        data = query.fetch()
        if not data:
            return None

        return data[0]


class Where:
    """Manages conditional filtering.

    Attributes:
        _criteria (list): The filtering criteria.
        _row (Row): The row to filter.
    """
    def __init__(self, row: "Row") -> None:
        """Initializes Where object.

        Args:
            row: The row to filter.
        """
        self._criteria = None
        self._row = row

    def criteria(self, column_name: str, operator: str, value: "Any") -> None:
        """Sets filtering criteria.

        Args:
            column_name: The column name to filter.
            operator: The operator sign.
            value: The value to compare.
        """
        self._criteria = [column_name, operator, value]

    def match(self) -> bool:
        """Returns whether conditions are met."""
        if not self._criteria:
            return False

        try:
            value = self._criteria[2]
            row_value = self._row[self._criteria[0]]
            match self._criteria[1]:
                case "=" | "eq" | "is":
                    return row_value == value
                case "!=" | "neq" | "is not":
                    return row_value != value
                case ">" | "gt":
                    return row_value > value
                case ">=" | "gte":
                    return row_value >= value
                case "<" | "lt":
                    return row_value < value
                case "<=" | "lte":
                    return row_value <= value
                case "in" | "is in":
                    return row_value in value
                case "not in":
                    return row_value not in value
                case "between":
                    return row_value >= value[0] and row_value <= value[1]
                case "not between":
                    return row_value <= value[0] and row_value >= value[1]
                case "like":
                    import re
                    pattern = re.escape(str(value))
                    if "%" not in pattern and "_" not in pattern:
                        return row_value == value

                    text = str(row_value)
                    pattern = re.sub(r"\%", ".*", pattern)
                    pattern = re.sub(r"\_", ".", pattern)
                    return re.search(pattern, text) is not None
                case "not like":
                    import re
                    pattern = re.escape(str(value))
                    if "%" not in pattern and "_" not in pattern:
                        return row_value != value

                    text = str(row_value)
                    pattern = re.sub(r"\%", ".*", pattern)
                    pattern = re.sub(r"\_", ".", pattern)
                    return re.search(pattern, text) is None
                case _:
                    return False
        except (TypeError, IndexError):
            return False

File: php/flintdb/Query.php
<?php

namespace FlintDB;

use DomainException,
    LogicException,
    ReflectionFunction,
    ReflectionMethod;

class Query {
  /**
   * The query content
   * @var array
   */
  private array $query = array();
  
  /**
   * To allow caching query result
   * @var bool
   */
  private bool $cache = true;
  
  /**
   * To detect whether query has been normalized
   * @var bool
   */
  private bool $ready = false;
  
  /**
   * Table instance to query data from
   * @var Table|null
   */
  private ?Table $table = null;
  
  /**
   * Parent Database instance of table
   * @var Database
   */
  private Database $database;
  
  /**
   * To slice query result
   * @var array
   */
  private array $limit = [ 0, -1 ];
  
  /**
   * Constructor
   * @param Database $database
   */
  public function __construct( Database $database ) {
    $this->database = $database;
  }
  
  /**
   * Sets the table to query data from
   * @param string $table
   * @return self
   */
  public function from( string $table ): self {
    $this->table = $this->database->table( $table );
    return $this;
  }
  
  /**
   * Performs left outer join
   * @param string $table
   * @param array $on
   * @param string|null $prefix
   * @return self
   */
  public function join( string $table, array $on, ?string $prefix = null ): self {
    $prefix = $prefix ?? $table . '.';
    $this->query[ 'join' ][ $table ] = [ $on, $prefix ];
    return $this;
  }
  
  /**
   * Maps callback to rows
   * @param array|string $callback
   * @return self
   */
  public function map( array | string $callback ): self {
    $reflection = is_array( $callback )
                    ? new ReflectionMethod( ...$callback )
                    : new ReflectionFunction( $callback );
    $this->query[ 'map' ][] = [ $callback, (string) $reflection ];
    return $this;
  }
  
  /**
   * Conditional filtering
   * @param string $column
   * @param string $operator
   * @param mixed $value
   * @return self
   */
  public function where( string $column, string $operator, mixed $value ): self {
    $this->query[ 'where' ][ $column ] = [ $operator, $value ];
    return $this;
  }
  
  /**
   * Renames column
   * @param string $column
   * @param string $new_name
   * @return self
   */
  public function select( string $column, string $new_name ): self {
    $this->query[ 'select' ][ $column ] = $new_name;
    return $this;
  }
  
  /**
   * Retrieves distinct, non repeated values
   * @param string $column
   * @return self
   */
  public function distinct( string $column ): self {
    $this->query[ 'distinct' ][] = $column;
    return $this;
  }
  
  /**
   * Sort rows base on column's value
   * @param string $column
   * @param string $order
   * @return self
   * @throws DomainException
   */
  public function sort( string $column, string $order = 'ASC' ): self {
    $order = strtoupper( $order );
    if ( 'ASC' !== $order && 'DESC' !== $order ) {
      throw new DomainException( 'Order must be either "ASC" or "DESC"' );
    }
    
    $this->query[ 'sort' ][ $column ] = $order === 'ASC' ? 1 : -1;
    return $this;
  }
  
  /**
   * Conditional filtering by mapping callback
   * @param array|string $callback
   * @return self
   */
  public function filter( array | string $callback ): self {
    $reflection = is_array( $callback )
                    ? new ReflectionMethod( ...$callback )
                    : new ReflectionFunction( $callback );
    $this->query[ 'filter' ][] = [ $callback, (string) $reflection ];
    return $this;
  }
  
  /**
   * Restrict the number of rows returned
   * @param int $max
   * @param int $offset
   * @return self
   * @throws DomainException
   */
  public function limit( int $max, int $offset = 0 ): self {
    if ( $max < 1 ) {
      throw new DomainException( 'Maximum limit must be greater than zero' );
    }
    
    $this->limit = [ $offset, $max ];
    return $this;
  }
  
  /**
   * Disable query result caching
   * @return self
   */
  public function no_cache(): self {
    $this->cache = false;
    return $this;
  }
  
  /**
   * Returns query result
   * @param int|null $cache_lifetime
   * @return Collection
   */
  public function fetch( ?int $cache_lifetime = null ): Collection {
    $this->normalize();
    if ( $this->cache ) {
      $cache = new Cache( ...[ ...$this->identifier(), $cache_lifetime ] );
      if ( $cache->valid() ) {
        return ( new Collection( $cache->get() ) )->limit( ...$this->limit );
      }
    }
    
    $data = array();
    foreach ( $this->table->rows() as $index => $row ) {
      if ( ! empty( $this->query[ 'join' ] ) ) {
        $join_object = new Join( $row );
        foreach ( $this->query[ 'join' ] as $table => $options ) {
          $join_object->table( $this->database->table( $table ) );
          $join_object->options( $options[0] );
          $join_row = $join_object->row();
          if ( $join_row ) {
            foreach ( $join_row->yield_columns() as $column => $value ) {
              $row[ $options[1] . $column ] = $value;
            }
          }
        }
      }
      
      foreach ( $this->query[ 'map' ] as $map ) {
        $map[0]( $row );
      }
      
      if ( ! empty( $this->query[ 'where' ] ) ) {
        $where_object = new Where( $row );
        foreach ( $this->query[ 'where' ] as $column => $criteria ) {
          $where_object->criteria( $column, ...$criteria );
          if ( $where_object->match() ) {
            $data[ $index ] = $row;
          }
        }
      } else {
        $data[ $index ] = $row;
      }
      
      if ( ! isset( $data[ $index ] ) ) {
        continue;
      }
      
      foreach ( $this->query[ 'select' ] as $column => $new_name ) {
        $row->rename_column( $column, $new_name );
      }
    }
    
    foreach ( $this->query[ 'distinct' ] as $column ) {
      $values = [];
      foreach ( $data as $index => $row ) {
        $value = $row[ $column ];
        if ( in_array( $value, $values, true ) ) {
          unset( $data[ $index ] );
        } else {
          $values[] = $value;
        }
      }
    }
    
    if ( ! empty( $this->query[ 'sort' ] ) ) {
      usort( $data, function ( $a, $b ) {
        foreach ( $this->query[ 'sort' ] as $column => $order ) {
          $cmp = match ( gettype( $a[ $column ] ) ) {
            'boolean' => (int) $a[ $column ] - (int) $b[ $column ],
            'double', 'integer' => ( $a[ $column ] > $b[ $column ] ) - ( $a[ $column ] < $b[ $column ] ),
            'string' => strcmp( $a[ $column ], $b[ $column ] ),
            default => $a[ $column ] <=> $b[ $column ]
          };
          
          if ( $cmp !== 0 ) {
            return $cmp * $order;
          }
        }
        
        return 0;
      });
    }
    
    foreach ( $this->query[ 'filter' ] as $filter ) {
      foreach ( $data as $index => $row ) {
        if ( ! $filter[0]( $row ) ) {
          unset( $data[ $index ] );
        }
      }
    }
    
    $data = array_values( $data );
    if ( $this->cache ) {
      $cache->put( $data );
    }
    
    return ( new Collection( $data ) )->limit( ...$this->limit );
  }
  
  /**
   * Normalizes query content
   * @return void
   * @throws LogicException
   */
  private function normalize(): void {
    if ( ! $this->table ) {
      throw new LogicException( 'Table must be specified' );
    }
    
    if ( $this->ready ) {
      return;
    }
    
    $this->query[ 'join' ] = $this->query[ 'join' ] ?? [];
    ksort( $this->query[ 'join' ] );
    
    $this->query[ 'map' ] = $this->query[ 'map' ] ?? [];
    sort( $this->query[ 'map' ] );
    
    $this->query[ 'where' ] = $this->query[ 'where' ] ?? [];
    ksort( $this->query[ 'where' ] );
    
    $this->query[ 'select' ] = $this->query[ 'select' ] ?? [];
    ksort( $this->query[ 'select' ] );
    
    $this->query[ 'distinct' ] = $this->query[ 'distinct' ] ?? [];
    sort( $this->query[ 'distinct' ] );
    
    $this->query[ 'sort' ] = $this->query[ 'sort' ] ?? [];
    ksort( $this->query[ 'sort' ] );
    
    $this->query[ 'filter' ] = $this->query[ 'filter' ] ?? [];
    sort( $this->query[ 'filter' ] );
    
    ksort( $this->query );
    $this->ready = true;
  }
  
  /**
   * Returns table and query content
   * @return array
   */
  public function identifier(): array {
    return [ $this->table, $this->query ];
  }
  
  /**
   * Resets query instance
   * @return self
   */
  public function clear(): self {
    $this->query = array();
    $this->cache = true;
    $this->ready = false;
    $this->limit = [ 0, -1 ];
    return $this;
  }
}
